include/albatross/src/linalg/block_diagonal.hpp

```diff
@@ -74,12 +93,12 @@ struct BlockDiagonal {
 /*
  * BlockDiagonalLDLT
  */
-template <class _Scalar, int _Rows, int _Cols>
-inline Eigen::Matrix<_Scalar, _Rows, _Cols> BlockDiagonalLDLT::solve(
-    const Eigen::Matrix<_Scalar, _Rows, _Cols> &rhs) const {
+template <typename Derived>
+inline Eigen::MatrixXd
```

Comment:
There must be a way to do this via Eigen types such that if the input is static sized the output will be too? `Eigen::Matrix<_Scalar, _Rows, _Cols>` would have allowed for that, but `Eigen::MatrixXd` will force all output to be dynamic doubles. I don't think it's a problem for us, and feel free to proceed without changing this, it just feels like there should be a way, `Eigen::DenseBase<Derived>::PlainMatrix` or something like that?

Reply:
I think the problem is not all expression types involved here derive from even `DenseBase`.

Reply:
`MatrixBase` inherits from `DenseBase` so it should right? https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/MatrixBase.h#L52

Reply:
Yes, but this gets used with things other than a `MatrixBase`.

Reply:
I guess I meant that if the input argument is a `MatrixBase<Derived>` then maybe you could get the `PlainMatrix` from it?

I doubt this would work but if `Eigen::MatrixBase<Derived>` works, then `Eigen::DenseBase<Derived>` should right?
```
template <typename Derived>
Eigen::DenseBase<Derived>::PlainMatrix BlockDiagonalLDLT::solve(const Eigen::MatrixBase<Derived> &rhs) const {
```

Anyway everytime I dig into Eigen I start to think I understand how to extend it, then I find something like `Solve.h` and get thoroughly lost. This isn't important enough to spend the time figuring it out.

Reply:
Sorry . . . I think the long-term "right way" is to make our `solve()` members look just like Eigen's ones and only overload `_solve_impl()` or whatever it is that takes an `RhsType` reference.